fpga_interchange/lookahead.cc

```diff
@@ -0,0 +1,1195 @@
+    xys.reserve(ctx->chip_info->tiles.size());
+
+    for (int32_t tile_type = 0; tile_type < ctx->chip_info->tile_types.ssize(); ++tile_type) {
+        indicies.clear();
+        xys.clear();
+
+        for (size_t tile = 0; tile < ctx->chip_info->tiles.size(); ++tile) {
+            if (ctx->chip_info->tiles[tile].type != tile_type) {
+                continue;
+            }
+
+            std::pair<size_t, size_t> xy;
+            ctx->get_tile_x_y(tile, &xy.first, &xy.second);
+
+            indicies.push_back(tile);
+            xys.push_back(xy);
+        }
+
+        auto &tile_sampler = tiles_of_type[tile_type];
+        tile_sampler.divide_samples(kNumberSamples, xys);
+
+        // Remap Sampler::indicies from 0 to number of tiles of type to
+        // absolute tile indicies.
+        for (size_t i = 0; i < indicies.size(); ++i) {
+            tile_sampler.indicies[i] = indicies[tile_sampler.indicies[i]];
+        }
+    }
+
+    if (ctx->verbose) {
+        log_info("Expanding input site wires\n");
+    }
+
+    // Expand backwards from each input site wire to find the cheapest
+    // non-site wire.
+    for (auto &input_pair : input_site_wires) {
+        expand_input_type(ctx, rng, tiles_of_type[input_pair.first.type], input_pair.first, &input_pair.second);
+    }
+
+    if (ctx->verbose) {
+        log_info("Expanding output site wires\n");
+    }
+
+    // Expand forward from each output site wire to find the cheapest
+    // non-site wire.
+    for (auto &output_pair : output_site_wires) {
+        output_pair.second.cost = std::numeric_limits<delay_t>::max();
+        expand_output_type(ctx, rng, tiles_of_type[output_pair.first.type], output_pair.first, &output_pair.second,
+                           &site_to_site_cost);
+    }
+    for (TypeWireId input_site_port : input_site_ports) {
+        expand_output_type(ctx, rng, tiles_of_type[input_site_port.type], input_site_port, nullptr, &site_to_site_cost);
+    }
+
+    if (ctx->verbose) {
+        log_info("Expanding all wire types\n");
+    }
+
+    DelayStorage all_tiles_storage;
+    all_tiles_storage.max_explore_depth = kInitialExploreDepth;
+
+    // These are wire types that have been explored.
+    absl::flat_hash_set<TypeWireSet> types_explored;
+
+    // These are wire types that have been deferred because they are trival
+    // copies of another wire type.  These can be cheaply computed after the
+    // graph has been explored.
+    absl::flat_hash_set<TypeWireId> types_deferred;
+
+    // Walk each tile type, and expand all non-site wires in the tile.
+    // Wires that are nodes will expand as if the node type is the first node
+    // in the wire.
+    //
+    // Wires that only have 1 output pip are deferred until the next loop,
+    // because generally those wires will get explored via another wire.
+    // The deferred will be expanded if this assumption doesn't hold.
+#if defined(NEXTPNR_USE_TBB) // Run parallely
```

Comment:
the slightly intricate use of macros here is going to make it very easily to accidentally break one of the TBB/non-TBB builds, I'm not sure if there is a good solution for this and ideally we test both in CI eventually

Reply:
I should refactor this to minimize the differences between the `#if defined` stuff

Reply:
I've refactored this so that most code compiles in both paths, and the differences are small and fairly local.